integration-tests/zombienet/src/lib.rs

```diff
@@ -0,0 +1,66 @@
+use anyhow::anyhow;
+use subxt::{OnlineClient, PolkadotConfig};
+use zombienet_sdk::{NetworkConfig, NetworkConfigBuilder, NetworkNode};
+
+pub mod environment;
+
+pub type Error = Box<dyn std::error::Error>;
+
+const CMD_TPL: &str = "chain-spec-generator {{chainName}}";
+
+pub fn small_network() -> Result<NetworkConfig, Error> {
+	let images = environment::get_images_from_env();
+	let config = NetworkConfigBuilder::new()
+		.with_relaychain(|r| {
+			r.with_chain("polkadot-local")
+				.with_default_command("polkadot")
+				.with_default_image(images.polkadot.as_str())
+				.with_chain_spec_command(CMD_TPL)
+				.chain_spec_command_is_local(true)
+				.with_node(|node| node.with_name("alice"))
+				.with_node(|node| node.with_name("bob"))
+		})
+		.with_parachain(|p| {
+			p.with_id(2000).cumulus_based(true).with_collator(|n| {
+				n.with_name("collator")
+					.with_command("polkadot-parachain")
+					.with_image(images.cumulus.as_str())
+			})
+		})
+		.build()
+		.map_err(|errs| {
+			let e = errs.iter().fold("".to_string(), |memo, err| format!("{memo} \n {err}"));
+			anyhow!(e)
+		})?;
+
+	Ok(config)
+}
+
+pub async fn wait_subxt_client(
+	node: &NetworkNode,
+) -> Result<subxt::OnlineClient<PolkadotConfig>, Error> {
+	log::debug!("trying to connect to: {}", node.ws_uri());
+	loop {
+		let res: Result<OnlineClient<PolkadotConfig>, Error> =
+			match node.client::<subxt::PolkadotConfig>().await {
+				Ok(cli) => {
+					break Ok(cli);
+				},
+				Err(e) => {
+					let cause = e.to_string();
+					log::trace!("{:?}", e);
+					if let subxt::Error::Rpc(subxt::error::RpcError::ClientError(inner)) = e {
+						log::trace!("inner: {}", inner.to_string());
+						if inner.to_string().contains("i/o error") {
```

Comment:
Yes, I'm not super happy with this. Maybe @jsdw can help with a better approach to check if the inner error is related to i/o.

Thx!

Reply:
Because somebody can plug in an arbitrary RPC client, we end up with this `Box<dyn std::error::Error + Send + Sync + 'static>` type to store errors from them. If you know which RPC client is used, you can try to cast that into the relevant error.

In the usual case, the error will be a [`jsonrpsee_core::ClientError`](https://docs.rs/jsonrpsee-core/0.22.5/jsonrpsee_core/client/error/enum.Error.html), so in that case you could do something like:

```rust
if let Ok(client_error) = inner.downcast::<jsonrpsee_core::ClientError> {
    if matches!(client_error, ClientError::Transport(_)) {
        // ...
    }
}
```

It's not much less ugly though in the end!

Not sure if this applies for you, but in Subxt the way that we wait for the node to be ready for connections is to wait for it to log that it's listening for them, ie https://github.com/paritytech/subxt/blob/ec7cea0acb59cdffb281aacd86a62c33514331e6/testing/substrate-runner/src/lib.rs#L191-L198 (that waits for a couple of "legacy" strings that nodes don't emit now too, which we can prob remove at this point).